src/Microsoft.DotNet.Cli.Utils/CommandResolver.cs

```diff
@@ -13,265 +13,54 @@ namespace Microsoft.DotNet.Cli.Utils
+                _defaultCommandResolver = DefaultCommandResolver.Create();
             }
 
-            var escapedArgs = ArgumentEscaper.EscapeAndConcatenateArgArrayForProcessStart(args);
-            return new CommandSpec(fileName, escapedArgs, CommandResolutionStrategy.NugetPackage);
+            return _defaultCommandResolver.Resolve(commandResolverArgs);
         }
-
-        private static CommandSpec CreateCommandSpecPreferringExe(
-            string commandName,
-            IEnumerable<string> args,
-            string commandPath,
-            CommandResolutionStrategy resolutionStrategy)
-        {
-            var useComSpec = false;
-            
-            if (PlatformServices.Default.Runtime.OperatingSystemPlatform == Platform.Windows &&
-                Path.GetExtension(commandPath).Equals(".cmd", StringComparison.OrdinalIgnoreCase))
-            {
-                var preferredCommandPath = Env.GetCommandPath(commandName, ".exe");
-
-                // Use cmd if we can't find an exe
-                if (preferredCommandPath == null)
-                {
-                    useComSpec = true;
-                }
-                else
-                {
-                    commandPath = preferredCommandPath;
-                }
-            }
-
-            if (useComSpec)
-            {
-                return CreateCmdCommandSpec(commandPath, args, resolutionStrategy);
-            }
-            else
-            {
-                var escapedArgs = ArgumentEscaper.EscapeAndConcatenateArgArrayForProcessStart(args);
-                return new CommandSpec(commandPath, escapedArgs, resolutionStrategy);
-            }
-        }
-
-        private static CommandSpec CreateCmdCommandSpec(
-            string command,
-            IEnumerable<string> args,
-            CommandResolutionStrategy resolutionStrategy)
+        
+        public static CommandSpec TryResolveScriptCommandSpec(
+            string commandName, 
+            IEnumerable<string> args, 
+            Project project, 
+            string[] inferredExtensionList)
         {
-            var comSpec = Environment.GetEnvironmentVariable("ComSpec");
-            
-            // Handle the case where ComSpec is already the command
-            if (command.Equals(comSpec, StringComparison.OrdinalIgnoreCase))
+            var commandResolverArgs = new CommandResolverArguments
             {
-                command = args.FirstOrDefault();
-                args = args.Skip(1);
-            }
-            var cmdEscapedArgs = ArgumentEscaper.EscapeAndConcatenateArgArrayForCmdProcessStart(args);
+                CommandName = commandName,
+                CommandArguments = args,
+                ProjectDirectory = project.ProjectDirectory,
+                InferredExtensions = inferredExtensionList
+            };
 
-            if (ArgumentEscaper.ShouldSurroundWithQuotes(command))
+            if (_scriptCommandResolver == null)
```

Comment:
Save on those allocations :smile: 

This is low cost so I've got nothing against just initializing every time.